ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/remote/MLSdkAsyncHttpResponseHandler.java

```diff
@@ -0,0 +1,194 @@
+                    ModelTensors tensors = processOutput(body, connector, scriptService, parameters, mlGuard);
+                    tensors.setStatusCode(statusCode);
+                    tensorOutputs.put(executionContext.getSequence(), tensors);
+                } catch (Exception e) {
+                    log.error("Failed to process response body: {}", body, e);
+                    if (executionContext.getExceptionHolder().get() == null)
+                        executionContext
+                            .getExceptionHolder()
+                            .compareAndSet(null, new MLException("Fail to execute predict in aws connector", e));
+                }
+            }
+        }
+    }
+
+    // Only all requests successful case will be processed here.
+    private void reOrderTensorResponses(Map<Integer, ModelTensors> tensorOutputs) {
+        ModelTensors[] modelTensors = new ModelTensors[tensorOutputs.size()];
+        log.debug("Reordered tensor outputs size is {}", tensorOutputs.size());
+        for (Map.Entry<Integer, ModelTensors> entry : tensorOutputs.entrySet()) {
+            modelTensors[entry.getKey()] = entry.getValue();
+        }
+        actionListener.onResponse(Arrays.asList(modelTensors));
+    }
+
+    protected class MLResponseSubscriber implements Subscriber<ByteBuffer> {
+        private Subscription subscription;
+
+        @Override
+        public void onSubscribe(Subscription s) {
+            this.subscription = s;
+            s.request(Long.MAX_VALUE);
+        }
+
+        @Override
+        public void onNext(ByteBuffer byteBuffer) {
+            responseBody.append(StandardCharsets.UTF_8.decode(byteBuffer));
+            subscription.request(Long.MAX_VALUE);
+        }
+
+        @Override
+        public void onError(Throwable t) {
+            log
+                .error(
+                    "Error on receiving response body from remote: {}",
+                    t instanceof NullPointerException ? "NullPointerException" : t.getMessage(),
+                    t
+                );
+            response(tensorOutputs);
+        }
+
+        @Override
+        public void onComplete() {
+            response(tensorOutputs);
+        }
+    }
+
+    private void response(Map<Integer, ModelTensors> tensors) {
+        processResponse(statusCode, responseBody.toString(), parameters, tensorOutputs);
+        executionContext.getCountDownLatch().countDown();
+        // when countdown's count equals to 0 means all responses are received.
+        if (executionContext.getCountDownLatch().getCount() == 0) {
+            if (executionContext.getExceptionHolder().get() != null) {
+                actionListener.onFailure(executionContext.getExceptionHolder().get());
+                return;
+            }
+            reOrderTensorResponses(tensors);
+        } else {
+            log.debug("Not all responses received, left response count is: " + executionContext.getCountDownLatch().getCount());
```

Comment:
I think we should return a failure for this case 
```
actionListener.onFailure(...)
```

Reply:
For this edge case, if we don't have all response back, we should still check if any failure received and return that failure; if no failure, return a general error message

Reply:
httpclient has timeout parameters by default, I don't see the case don't have all response back. Can you give an example?

Reply:
I don't have example on mind. Generally we use a expected way to process this. I think it's expected to call onFailure for such error, rather than wait for the timeout exception happens.

Reply:
Under the condition that "executionContext.getCountDownLatch().getCount() != 0 and executionContext.getExceptionHolder().get() != null", the handler would still wait for the next responses. This means that when we receive an Exception in between the first and last response, we still wait for the last response to return the error message from the earlier results. This is a wast of time and we should just return a failure in the short circuit way and stop processing the remaining responses.